File: src/main/java/club/mindtech/mindbot/commands/BaseCommand.java
package club.mindtech.mindbot.commands;

import club.mindtech.mindbot.util.StrUtil;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.commands.build.Commands;
import net.dv8tion.jda.api.interactions.commands.build.SlashCommandData;

public abstract class BaseCommand {
    private final String name;
    private final String description;
    private final String usage;

    public BaseCommand(String name, String description, String usage) {
        this.name = name;
        this.description = description;
        this.usage = StrUtil.prefix(usage);
    }

    public String getName() {
        return name;
    }

    public String getDescription() {
        return description;
    }

    public String getUsage() {
        return usage;
    }

    public SlashCommandData getCommandData() {
        return Commands.slash(name, description);
    }

    public void onSlashCommand(SlashCommandInteractionEvent event) {}
}

File: src/main/kotlin/club/mindtech/mindbot/commands/BaseCommand.kt
package club.mindtech.mindbot.commands

import net.dv8tion.jda.api.events.interaction.command.CommandAutoCompleteInteractionEvent
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent
import net.dv8tion.jda.api.events.interaction.component.SelectMenuInteractionEvent
import net.dv8tion.jda.api.interactions.commands.build.Commands
import net.dv8tion.jda.api.interactions.commands.build.SlashCommandData

const val ID_SEPARATOR = ":"

abstract class BaseCommand(val name: String, val description: String, usage: String) {
    val usage: String

    init {
        this.usage = "/$usage"
    }

    fun generateClassId(vararg ids: String): String = listOf(name, *ids).joinToString(ID_SEPARATOR)

    open fun getCommandData(): SlashCommandData = Commands.slash(name, description)

    open fun onSlashCommand(event: SlashCommandInteractionEvent) {}

    open fun onAutoComplete(event: CommandAutoCompleteInteractionEvent) {}

    open fun onButtonInteraction(event: ButtonInteractionEvent, idArgs: List<String>) {}

    open fun onMenuInteraction(event: SelectMenuInteractionEvent, idArgs: List<String>) {}
}
